cmogstored.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.38.2.
.TH CMOGSTORED "1" "March 2013" "cmogstored 1.2.1" "User Commands"
.SH NAME
cmogstored \- manual page for cmogstored 1.2.1
.SH SYNOPSIS
.B cmogstored
[\fIOPTION\fR...]
.SH DESCRIPTION
cmogstored \fB\-\-\fR alternative mogstored implementation for MogileFS
.TP
\fB\-\-config=\fR<file>
Set config file (no default, unlike mogstored)
.TP
\fB\-d\fR, \fB\-\-daemonize\fR
Daemonize
.TP
\fB\-\-docroot=\fR<path>
Docroot above device mount points.  Defaults to
/var/mogdata
.HP
\fB\-\-httplisten=\fR<ip:port> IP/Port HTTP server listens on
.TP
\fB\-\-maxconns=\fR<number>
Number of simultaneous clients to serve. Default
10000
.HP
\fB\-\-mgmtlisten=\fR<ip:port> IP/Port management/sidechannel listens on
.TP
\fB\-\-pidfile=\fR<path>
path to PID file
.TP
\-?, \fB\-\-help\fR
give this help list
.TP
\fB\-\-usage\fR
give a short usage message
.TP
\fB\-V\fR, \fB\-\-version\fR
print program version
.SH ENVIRONMENT
MOG_IOSTAT_CMD - command-line for invoking iostat(1).  This is used
by the sidechannel to report disk utilization to trackers.
Default: "iostat -dx 1 30"
.SH SIGNALS
SIGQUIT - gracefully shutdown the server
SIGUSR2 - upgrade executable on the fly
.SH UPGRADING
Upgrading cmogstored on-the-fly is possible since version 1.2.0.
The procedure is close to the one used by nginx users: send SIGUSR2
followed by SIGQUIT to the old process.  Use of SIGWINCH and SIGHUP
are currently not supported (but may be in the future).
Procedures for the nginx upgrade are documented here:
http://wiki.nginx.org/CommandLine#Upgrading_To_a_New_Binary_On_The_Fly
.SH "REPORTING BUGS"
Only bug reports in plain-text email will be read.
You may also report bugs publically to the MogileFS mailing list
mogile@googlegroups.com.
.PP
.br
Report bugs to normalperson@yhbt.net.
.SH COPYRIGHT
Copyright (C) 2012-2013, Eric Wong <normalperson@yhbt.net>.
License: GNU GPL version 3 or later <https://www.gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
cmogstored website: http://bogomips.org/cmogstored/README
You can learn more about MogileFS: http://mogilefsd.org/

cmogstored source code is available via git:

	git clone git://bogomips.org/cmogstored.git


NEWS
1.2.1 / 2013-03-04 01:33 UTC
----------------------------

  This release only fixes an assertion failure during graceful shutdown
  while MogileFS fsck is running with checksumming enabled.

  This only affects users running fsck with checksumming enabled during a
  graceful shutdown of cmogstored.  For upgrading cmogstored it is
  recommended to:

  1) stop fsck on the trackers (via "mogadm fsck stop")
  2) wait for all tracker queues to drain and stop sending
     fsck traffic to the affected host.  You may wish to
     "!want 0 fsck" on all your trackers and wait for the
     fsck workers to stop.
  3) upgrade cmogstored (in place upgrade works)

  There are also several code comment updates for internal
  components of cmogstored which may interest potential hackers.

1.2.0 / 2013-02-18 23:39 UTC
----------------------------

  This release suppors nginx-style binary upgrades via SIGUSR2.
  The behavior of this process should match that of nginx:
  http://wiki.nginx.org/CommandLine#Upgrading_To_a_New_Binary_On_The_Fly
  SIGWINCH and SIGHUP are currently no-ops and may match nginx behavior
  in the future.  They are not required for binary upgrades.

  Slow/unreliable mount points (if you have them) should have less effect
  on iostat sidechannel clients once the process is running.  Startup is
  still slow with unreliable mount points, unfortunately.

  Error handling is now graceful for thread creation failures and
  systems lacking stdio memstream support (FreeBSD).

1.1.0 / 2013-01-18 10:54 UTC
----------------------------

  cmogstored now works around an EPOLL_CTL_MOD race condition
  in old kernels.  This workaround is unneeded and disabled on
  Linux v3.0.59+, v3.2.37+, v3.4.26+, v3.5.7.3+, v3.7.3+ and v3.8+

  FreeBSD users should no longer see ECONNRESET errors on
  close(2).

  Unnecessary mkdir/mkdirat syscalls are optimistically avoided on PUT.
  MSG_MORE is used (instead of TCP_CORK) on Linux to avoid extra
  syscalls.  We also avoid POSIX_FADV_SEQUENTIAL when returning
  small responses.

1.0.0 / 2012-12-13 08:38 UTC
----------------------------

  FreeBSD support is improved:
  - ZFS support (tested on FreeBSD 9.0)
  - Fix disk usage report on FSes w/ small fragment size (ZFS)
  - faster graceful shutdown on kqueue()-based systems.
  - systems lacking open_memstream() no longer misreports OOM on SIGQUIT

  Thanks to Ask Bjørn Hansen for helping with the ZFS support

  Linux:
  - several bugfixes for handling of out-of-FD situations

  There are also minor documentation and build improvements.

0.9.0 / 2012-12-05 01:15 UTC
----------------------------

  This release handles out-of-disk-space errors from PUT and usage
  file generation more gracefully.  Failed PUTs due to client-side
  disconnects are also logged with additional debugging information.
  There are also minor internal cleanups for shutdowns.

0.8.0 / 2012-11-14 21:34 UTC
----------------------------

  HTTP connections remain persistent after a failed Content-MD5, this
  prevents the MogileFS monitor from generating excessive TIME-WAIT
  connections.

  Linux only: Idle HTTP connections are automatically closed under FD
  pressure from systems with too many trackers.  This adds zero
  overhead unless the process runs out of file descriptors.

  "server aio_threads = <digit>" is supported and attempts to mimic the
  Perlbal interface.  There is no error reporting nor reporting of the
  current thread count.  Requests are silently capped in the 1-100
  (inclusive) range.

  The "shutdown" command from Perlbal is also supported.

  The Content-Range response header (for partial requests) was missing a
  '\r' and thus not compliant with extremely strict HTTP parsers.

0.7.0 / 2012-10-17 00:06 UTC
----------------------------

  This release fixes a boundary error in Content-MD5 trailer
  reading+parsing when receiving chunked PUT requests of certain
  sizes.  This bug was found while attempting to upload a
  198689228 byte file using 16K chunks (using the Ruby
  mogilefs-client 3.4.0 to send the Content-MD5 trailer)

  There are also minor cleanups, new test cases and better error
  reporting for ENOSPC errors.

0.6.0 / 2012-10-04 09:05 UTC
----------------------------

  This release fixes a concurrency assertion failure affecting
  sidechannel+checksum users.  This bug has existed since 0.2.0, but was
  not noticed without the concurrency simplification in 0.5.0

  There are additional test cases for checksumming and gnulib updates
  for the tarball.

0.5.0 / 2012-09-10 01:14 UTC
----------------------------

  * I/O utilization is now reported correctly when multiple
    MogileFS devices share the same local filesystem

  * attempt to reduce client-side syscalls on low-latency networks
    with TCP_NOPUSH/TCP_CORK

  * remove a tiny chance of starvation when sharing work between threads
    while improving theoretical fairness

  * minor code cleanups and gnulib updates

  * acceptor threads push directly into the event queue to avoid
    poorly-defined TCP_DEFER_ACCEPT/accept filter semantics

  * acceptor threads no longer touch the FS, so they are now scaled
    to CPU count, not filesystem count

  * absolute URIs are now supported in HTTP/1.1 requests

0.4.0 / 2012-05-19 00:37 UTC
----------------------------

  * The kqueue code path now requires no more syscalls than the epoll
    code path on *BSDs.

  * avoids malloc() usage after fork() (for iostat(1)).  This works
    around some malloc() implementations which fail to reinitialize
    malloc locks after fork().

  * iostat(1) is no longer spawned for HTTP-only deployments.

  * usage files are no longer created in HTTP-only deployments.

  * FIONBIO + ioctl() is no longer required, we'll fall back to the
    slower (but POSIX) fcntl() equivalent if FIONBIO isn't available.

  * Testers without native epoll or kqueue support, the the current
    stable branch of libkqueue (r549) should work as should future
    releases.

cmogstored.c
/*
 * Copyright (C) 2012-2013, Eric Wong <normalperson@yhbt.net>
 * License: GPLv3 or later (see COPYING for details)
 */
#include "cmogstored.h"
#include "cfg.h"
#include "nostd/setproctitle.h"
#define THIS "cmogstored"
#define DESC "alternative mogstored implementation for MogileFS"
static char summary[] = THIS " -- "DESC;
const char *argp_program_bug_address = PACKAGE_BUGREPORT;
const char *argp_program_version = THIS" "PACKAGE_VERSION;
static struct mog_fd *master_selfwake;
static sig_atomic_t sigchld_hit;
static sig_atomic_t do_exit;
static sig_atomic_t do_upgrade;
static size_t nthr;
static bool have_mgmt;
static pid_t master_pid;
static pid_t upgrade_pid;
static unsigned long worker_processes;
static bool iostat_running;

#define CFG_KEY(f) -((int)offsetof(struct mog_cfg,f) + 1)
static struct argp_option options[] = {
	{ .name = "daemonize", .key = 'd',
	  .doc = "Daemonize" },
	{ .name = "config", .key = CFG_KEY(configfile),
	  .arg = "<file>",
          .doc = "Set config file (no default, unlike mogstored)" },
	{ .name = "httplisten", .key = CFG_KEY(httplisten),
	  .arg = "<ip:port>",
	  .doc = "IP/Port HTTP server listens on" },
	/*
	 * NOT adding httpgetlisten here to avoid (further) breaking
	 * compatibility with Perl mogstored.  Most of these command-line
	 * options suck anyways.
	 */
	{ .name = "mgmtlisten", .key = CFG_KEY(mgmtlisten),
	  .arg = "<ip:port>",
	  .doc = "IP/Port management/sidechannel listens on" },
	{ .name = "docroot", .key = CFG_KEY(docroot),
	  .arg = "<path>",
	  .doc = "Docroot above device mount points.  "
	         "Defaults to "MOG_DEFAULT_DOCROOT
	},
	{ .name = "maxconns", .key = CFG_KEY(maxconns),
	  .arg = "<number>",
	  .doc = "Number of simultaneous clients to serve. "
	         "Default " MOG_STR(MOG_DEFAULT_MAXCONNS) },
	{ .name = "pidfile", .key = CFG_KEY(pidfile),
	  .arg = "<path>",
	  .doc = "path to PID file" },
	{ .name = "server", .key = CFG_KEY(server), .flags = OPTION_HIDDEN,
	  .arg = "(perlbal|none)"
	},
	{
	  /* hidden for now, don't break compat with Perl mogstored */
	  .name = "multi", .key = -'M', .flags = OPTION_HIDDEN
	},
	{
	  /* hidden for now, don't break compat with Perl mogstored */
	  .name = "worker-processes", .key = -'W', .flags = OPTION_HIDDEN,
	  .arg = "<number>",
	},
	/* we don't load a default config file like Perl mogstored at all */
	{ .name = "skipconfig", .key = 's', .flags = OPTION_HIDDEN },
	{ NULL }
};

static void new_cfg_or_die(const char *config)
{
	struct mog_cfg *cfg = mog_cfg_new(config);

	if (!cfg) die("invalid (or duplicate) config=%s", config);
	if (mog_cfg_load(cfg) == 0) return;

	die("failed to load config=%s: %s",
	    config, errno ? strerror(errno) : "parser error");
}

static void check_strtoul(unsigned long *dst, const char *s, const char *key)
{
	char *end;

	errno = 0;
	*dst = strtoul(s, &end, 10);
	if (errno)
		die_errno("failed to parse --%s=%s", key, s);
	if (*end)
		die("failed to parse --%s=%s (invalid character: %c)",
		    key, s, *end);
}

static void addr_or_die(struct mog_addrinfo **dst, const char *key, char *s)
{
	*dst = mog_listen_parse(s);
	if (!*dst)
		die("failed to parse %s=%s", key, s);
}

static error_t parse_opt(int key, char *arg, struct argp_state *state)
{
	struct mog_cfg *cfg = state->input;
	int rv = 0;

	switch (key) {
	case 'd': cfg->daemonize = true; break;
	case 's': /* no-op, we don't load the default config file */; break;
	case CFG_KEY(docroot): cfg->docroot = xstrdup(arg); break;
	case CFG_KEY(pidfile): cfg->pidfile = xstrdup(arg); break;
	case CFG_KEY(configfile): new_cfg_or_die(arg); break;
	case CFG_KEY(maxconns):
		check_strtoul(&cfg->maxconns, arg, "maxconns");
		break;
	case CFG_KEY(httplisten):
		addr_or_die(&cfg->httplisten, "httplisten", arg);
		break;
	case CFG_KEY(mgmtlisten):
		addr_or_die(&cfg->mgmtlisten, "mgmtlisten", arg);
		break;
	case CFG_KEY(server):
		cfg->server = xstrdup(arg);
		mog_cfg_check_server(cfg);
		break;
	case -'M': mog_cfg_multi = true; break;
	case -'W':
		check_strtoul(&worker_processes, arg, "worker-processes");
		if (worker_processes > UINT_MAX)
			die("--worker-processes exceeded (max=%u)", UINT_MAX);
		break;
	case ARGP_KEY_ARG:
		argp_usage(state);
	case ARGP_KEY_END:
		break;
	default:
		rv = ARGP_ERR_UNKNOWN;
	}

	return rv;
}

static void dup2_null(int oldfd, int newfd, const char *errdest)
{
	int rc;

	do {
		rc = dup2(oldfd, newfd);
	} while (rc < 0 && (errno == EINTR || errno == EBUSY));

	if (rc < 0)
		die_errno("dup2(/dev/null,%s) failed", errdest);
}

static void daemonize(void)
{
	int ready_pipe[2];
	pid_t pid;
	ssize_t r;
	int fd;

	if (pipe(ready_pipe) < 0)
		die_errno("pipe() failed");
	pid = fork();
	if (pid < 0)
		die_errno("fork() failed");
	if (pid > 0) { /* parent */
		mog_close(ready_pipe[1]);
		do {
			r = read(ready_pipe[0], &pid, sizeof(pid_t));
		} while (r < 0 && errno == EINTR);

		PRESERVE_ERRNO( mog_close(ready_pipe[0]) );
		if (r == sizeof(pid_t))
			exit(EXIT_SUCCESS);
		if (r < 0)
			die_errno("ready_pipe read error");
		die("daemonization error, check syslog");
	}

	/* child */
	mog_close(ready_pipe[0]);
	if (setsid() < 0)
		die_errno("setsid() failed");

	pid = fork();
	if (pid < 0)
		die_errno("fork() failed");
	if (pid > 0) /* intermediate parent */
		exit(EXIT_SUCCESS);

	if (chdir("/") < 0)
		die_errno("chdir(/) failed");
	fd = open("/dev/null", O_RDWR);
	if (fd < 0)
		die_errno("open(/dev/null) failed");

	dup2_null(fd, STDIN_FILENO, "stdin");
	dup2_null(fd, STDOUT_FILENO, "stdout");
	dup2_null(fd, STDERR_FILENO, "stderr");
	mog_close(fd);

	do {
		r = write(ready_pipe[1], &pid, sizeof(pid_t));
	} while (r < 0 && errno == EINTR);
	if (r < 0)
		syslog(LOG_CRIT, "ready_pipe write error: %m");
	else
		assert(r == sizeof(pid_t) && "impossible short write");
	mog_close(ready_pipe[1]);
}

#ifndef LOG_PERROR
# define LOG_PERROR 0
#endif

/* TODO: make logging configurable (how?) */
static void log_init(bool is_daemon)
{
	int mask = 0;
	int option = LOG_PID;

	if (!is_daemon)
		option |= LOG_PERROR;

	openlog(THIS, option, LOG_DAEMON);
	mask |= LOG_MASK(LOG_EMERG);
	mask |= LOG_MASK(LOG_ALERT);
	mask |= LOG_MASK(LOG_CRIT);
	mask |= LOG_MASK(LOG_ERR);
	mask |= LOG_MASK(LOG_WARNING);
	mask |= LOG_MASK(LOG_NOTICE);
	mask |= LOG_MASK(LOG_INFO);
	/* mask |= LOG_MASK(LOG_DEBUG); */
	setlogmask(mask);
}

MOG_NOINLINE static void setup(int argc, char *argv[])
{
	int pid_fd = -1;
	static struct argp argp = { options, parse_opt, NULL, summary };

	mog_mnt_refresh();
	argp_parse(&argp, argc, argv, 0, NULL, &mog_cli);
	mog_cfg_validate_or_die(&mog_cli);
	log_init(mog_cli.daemonize);
	mog_inherit_init();
	mog_cfg_svc_start_or_die(&mog_cli);
	mog_inherit_cleanup();

	if (mog_cli.pidfile) pid_fd = mog_pidfile_prepare(mog_cli.pidfile);

	/* don't daemonize if we're inheriting FDs, we're already daemonized */
	if (mog_cli.daemonize && !getenv("CMOGSTORED_FD"))
		daemonize();

	if (pid_fd >= 0 && mog_pidfile_commit(pid_fd) < 0)
		syslog(LOG_ERR,
		       "failed to write pidfile(%s): %m. continuing...",
		       mog_cli.pidfile);

	master_pid = getpid();

	/* 10 - 100 threads based on number of devices, same as mogstored */
	nthr = mog_mkusage_all() * 10;
	nthr = MAX(10, nthr);
	nthr = MIN(100, nthr);
}

/* Hash iterator function */
static bool svc_start_each(void *svcptr, void *qptr)
{
	struct mog_svc *svc = svcptr;
	struct mog_queue *q = qptr;
	struct mog_accept *ac;
	size_t athr = (size_t)num_processors(NPROC_CURRENT);

	/*
	 * try to distribute accept() callers between workers more evenly
	 * with wake-one accept() behavior by trimming down on acceptors
	 */
	if (worker_processes) {
		athr /= worker_processes;
		if (athr == 0)
			athr = 1;
	}

	svc->queue = q;

	if (svc->mgmt_fd >= 0) {
		have_mgmt = true;
		ac = mog_accept_init(svc->mgmt_fd, svc, mog_mgmt_post_accept);

		/*
		 * mgmt port is rarely used and always persistent, so it
		 * does not need multiple threads for blocking accept()
		 */
		mog_thrpool_start(&ac->thrpool, 1, mog_accept_loop, ac);
	}

	if (svc->http_fd >= 0) {
		ac = mog_accept_init(svc->http_fd, svc, mog_http_post_accept);
		mog_thrpool_start(&ac->thrpool, athr, mog_accept_loop, ac);
	}

	if (svc->httpget_fd >= 0) {
		ac = mog_accept_init(svc->httpget_fd, svc,
		                     mog_httpget_post_accept);
		mog_thrpool_start(&ac->thrpool, athr, mog_accept_loop, ac);
	}

	return true;
}

static void worker_wakeup_handler(int signum)
{
	switch (signum) {
	case SIGUSR2: do_upgrade = 1; break;
	case SIGCHLD: sigchld_hit = 1; break;
	case SIGQUIT:
	case SIGTERM:
	case SIGINT:
		do_exit = signum;
	}
	mog_notify(MOG_NOTIFY_SIGNAL);
}

static void wakeup_noop(int signum)
{
	/* just something to cause EINTR */
}

static void master_wakeup_handler(int signum)
{
	switch (signum) {
	case SIGCHLD: sigchld_hit = 1; break;
	case SIGUSR2: do_upgrade = 1; break;
	case SIGQUIT:
	case SIGTERM:
	case SIGINT:
		do_exit = signum;
	}
	mog_selfwake_trigger(master_selfwake);
}

static void siginit(void (*wakeup_handler)(int))
{
	struct sigaction sa;

	memset(&sa, 0, sizeof(struct sigaction));
	CHECK(int, 0, sigemptyset(&sa.sa_mask) );

	sa.sa_handler = SIG_IGN;
	CHECK(int, 0, sigaction(SIGPIPE, &sa, NULL));

	sa.sa_handler = wakeup_noop;
	CHECK(int, 0, sigaction(SIGURG, &sa, NULL));

	sa.sa_handler = wakeup_handler;

	/* TERM and INT are graceful shutdown for now, no immediate shutdown */
	CHECK(int, 0, sigaction(SIGTERM, &sa, NULL));
	CHECK(int, 0, sigaction(SIGINT, &sa, NULL));

	CHECK(int, 0, sigaction(SIGQUIT, &sa, NULL)); /* graceful shutdown */
	CHECK(int, 0, sigaction(SIGUSR1, &sa, NULL)); /* no-op, nginx compat */
	CHECK(int, 0, sigaction(SIGUSR2, &sa, NULL)); /* upgrade */

	/*
	 * SIGWINCH/SIGHUP are no-ops for now to allow reuse of nginx init
	 * scripts.  We should support them in the future.
	 * SIGWINCH will disable new connections and drop idlers
	 * SIGHUP will reenable new connections/idlers after SIGWINCH
	 */
	CHECK(int, 0, sigaction(SIGWINCH, &sa, NULL));
	CHECK(int, 0, sigaction(SIGHUP, &sa, NULL));

	sa.sa_flags = SA_NOCLDSTOP;
	CHECK(int, 0, sigaction(SIGCHLD, &sa, NULL));
}

static void process_died(pid_t pid, int status);

static void sigchld_handler(void)
{
	sigchld_hit = 0;

	for (;;) {
		int status;
		pid_t pid = waitpid(-1, &status, WNOHANG);

		if (pid > 0) {
			process_died(pid, status);
		} else if (pid == 0) {
			return;
		} else {
			switch (errno) {
			case EINTR: sigchld_hit = 1; return; /* retry later */
			case ECHILD: return;
			default: die_errno("waitpid");
			}
		}
	}
}

static void upgrade_handler(void)
{
	do_upgrade = 0;
	if (upgrade_pid > 0) {
		syslog(LOG_INFO, "upgrade already running on PID:%d",
		       upgrade_pid);
	} else {
		if (master_pid == getpid())
			upgrade_pid = mog_upgrade_spawn();
		/* else: worker processes (if configured) do not upgrade */
	}
}

static void main_worker_loop(const pid_t parent)
{
	mog_cancel_disable(); /* mog_idleq_wait() now relies on this */
	while (parent == 0 || parent == getppid()) {
		mog_notify_wait(have_mgmt);
		if (sigchld_hit)
			sigchld_handler();
		if (do_upgrade)
			upgrade_handler();
		if (do_exit)
			cmogstored_exit();
		if (have_mgmt)
			mog_mnt_refresh();
		else if (have_mgmt && !iostat_running && !do_exit)
			/*
			 * maybe iostat was not installed/available/usable at
			 * startup, but became usable later
			 */
			iostat_running = mog_iostat_respawn(0);
	}

	syslog(LOG_INFO, "parent=%d abandoned us, dying", parent);
	cmogstored_exit();
}

static void run_worker(const pid_t parent)
{
	struct mog_queue *q = mog_queue_new();

	mog_notify_init();
	siginit(worker_wakeup_handler);
	mog_thrpool_start(&q->thrpool, nthr, mog_queue_loop, q);
	have_mgmt = false;
	mog_svc_each(svc_start_each, q); /* this will set have_mgmt */
	if (have_mgmt) {
		iostat_running = mog_iostat_respawn(0);
		if (!iostat_running)
			syslog(LOG_WARNING, "iostat(1) not available/running");
	}
	main_worker_loop(parent);
}

static void fork_worker(unsigned worker_id)
{
	pid_t pid;
	pid_t parent = getpid(); /* not using getppid() since it's racy */

	pid = fork();
	if (pid > 0) {
		mog_process_register(pid, worker_id);
	} else if (pid == 0) {
		/* workers have no workers of their own */
		worker_processes = 0;
		mog_process_reset();

		/* worker will call mog_intr_enable() later in notify loop */
		run_worker(parent);
		exit(EXIT_SUCCESS);
	} else {
		syslog(LOG_ERR, "fork() failed: %m, sleeping for 10s");
		mog_sleep(10);
	}
}

/* run when a worker dies */
static void worker_died(pid_t pid, int status, unsigned id)
{
	if (do_exit)
		return;

	syslog(LOG_INFO,
	       "worker[%u] PID:%d died with status=%d, respawning",
	       id, (int)pid, status);
	fork_worker(id);
}

static void iostat_died(pid_t pid, int status)
{
	if (do_exit)
		return;
	iostat_running = mog_iostat_respawn(status);
}

/* run when any process dies */
static void process_died(pid_t pid, int status)
{
	unsigned id = mog_process_reaped(pid);
	char *name;

	if (mog_process_is_worker(id)) {
		worker_died(pid, status, id);
		return;
	}

	switch (id) {
	case MOG_PROC_IOSTAT:
		assert(worker_processes == 0 &&
		       "master process registered iostat process");
		iostat_died(pid, status);
		return;
	case MOG_PROC_UPGRADE:
		assert(pid == upgrade_pid && "upgrade_pid misplaced");
		syslog(LOG_INFO, "upgrade PID:%d exited with status=%d",
		       pid, status);
		mog_pidfile_upgrade_abort();
		upgrade_pid = -1;
		return;
	default:
		/* could be an inherited iostat if we're using worker+master */
		name = mog_process_name(id);
		syslog(LOG_INFO,
		       "reaped %s pid=%d with status=%d, ignoring",
		       name, (int)pid, status);
		free(name);
	}
}

static void run_master(void)
{
	unsigned id;
	size_t running = worker_processes;

	master_selfwake = mog_selfwake_new();
	siginit(master_wakeup_handler);

	for (id = 0; id < worker_processes; id++)
		fork_worker(id);

	while (running > 0) {
		mog_selfwake_wait(master_selfwake);
		if (sigchld_hit)
			sigchld_handler();
		if (do_upgrade)
			upgrade_handler();
		if (do_exit)
			running = mog_kill_each_worker(SIGQUIT);
	}
}

int main(int argc, char *argv[], char *envp[])
{
	mog_upgrade_prepare(argc, argv, envp);
	/* hack for older gcov + gcc, see nostd/setproctitle.h */
	spt_init(argc, argv, envp);
	set_program_name(argv[0]);

	mog_intr_disable();
	setup(argc, argv); /* this daemonizes */

	mog_process_init(worker_processes);
	if (worker_processes == 0)
		run_worker(0);
	else
		run_master();

	return 0;
}

/* called by the "shutdown" command via mgmt */
void cmogstored_quit(void)
{
	if (master_pid != getpid()) {
		if (kill(master_pid, SIGQUIT) != 0)
			syslog(LOG_ERR,
			       "SIGQUIT failed on master process (pid=%d): %m",
				master_pid);
	} else {
		worker_wakeup_handler(SIGQUIT);
	}
}
